src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathV2Spec.groovy

```diff
@@ -448,14 +449,92 @@ class ProtectedPathV2Spec extends HealthCheckSpecification {
             }
         }
 
+        when: "ISL gets back up"
+        def mainIslUp = antiflap.portUp(mainIsl.srcSwitch.dpId, mainIsl.srcPort)
+        Wrappers.wait(WAIT_OFFSET) { assert northbound.getLink(mainIsl).state == IslChangeType.DISCOVERED }
+
+        then: "Main path remains the same, flow becomes UP, main path UP, protected UP"
+        Wrappers.wait(WAIT_OFFSET) {
+            def newPath = northbound.getFlowPath(flow.flowId)
+            assert pathHelper.convert(newPath) == pathHelper.convert(path.protectedPath)
+            verifyAll(northbound.getFlow(flow.flowId)) {
+                status == FlowState.UP.toString()
+                flowStatusDetails.mainFlowPathStatus == "Up"
+                flowStatusDetails.protectedFlowPathStatus == "Up"
+            }
+        }
+
         cleanup:
         flowHelperV2.deleteFlow(flow.flowId)
-        mainIsl && antiflap.portUp(mainIsl.srcSwitch.dpId, mainIsl.srcPort)
+        mainIslDown && !mainIslUp && antiflap.portUp(mainIsl.srcSwitch.dpId, mainIsl.srcPort)
         otherIsls && otherIsls.collectMany{[it, it.reversed]}.each { database.resetIslBandwidth(it) }
         Wrappers.wait(WAIT_OFFSET) { assert northbound.getLink(mainIsl).state == IslChangeType.DISCOVERED }
         database.resetCosts()
     }
 
+    @Tidy
+    @Ignore("https://github.com/telstra/open-kilda/issues/4038")
+    def "Flow swaps to protected path when main path gets broken, becomes DEGRADED if protected path is unable to reroute(no path)"() {
+        given: "Two switches with 2 diverse paths at least"
+        def switchPair = topologyHelper.switchPairs.find {
+            it.paths.unique(false) { a, b -> a.intersect(b) == [] ? 1 : 0 }.size() > 1
+        } ?: assumeTrue("No switches with at least 2 diverse paths", false)
+
+        when: "Create flow with protected path"
+        def flow = flowHelperV2.randomFlow(switchPair).tap { allocateProtectedPath = true }
+        flowHelperV2.addFlow(flow)
+        def path = northbound.getFlowPath(flow.flowId)
+
+        and: "Other paths are not available (ISLs are down)"
+        def usedIsls = pathHelper.getInvolvedIsls(pathHelper.convert(path.protectedPath)) +
+            pathHelper.getInvolvedIsls(pathHelper.convert(path))
```

Comment:
As for me would be greate to extract  `pathHelper.convert(path.protectedPath) / pathHelper.convert(path)` into the separate variables.

Reply:
added 
```
        def originalMainPath = pathHelper.convert(path)
        def originalProtectedPath = pathHelper.convert(path.protectedPath)
```
and refactored usages